server/src-lib/Hasura/Server/Migrate.hs

```diff
@@ -123,37 +128,91 @@ migrateCatalog migrationTime = do
 
     -- migrates an existing catalog to the latest version from an existing verion
     migrateFrom :: T.Text -> m MigrationResult
-    migrateFrom previousVersion
-      | previousVersion == latestCatalogVersionString = pure MRNothingToDo
-      | [] <- neededMigrations = throw400 NotSupported $
-          "Cannot use database previously used with a newer version of graphql-engine (expected"
-            <> " a catalog version <=" <> latestCatalogVersionString <> ", but the current version"
-            <> " is " <> previousVersion <> ")."
-      | otherwise =
-          traverse_ snd neededMigrations *> postMigrate $> MRMigrated previousVersion
+    migrateFrom previousVersion =
+      case downgradeOpts of
+        Nothing
+          | previousVersion == latestCatalogVersionString ->
+              pure MRNothingToDo
+          | [] <- neededMigrations ->
+              throw400 NotSupported $
+                "Cannot use database previously used with a newer version of graphql-engine (expected"
+                  <> " a catalog version <=" <> latestCatalogVersionString <> ", but the current version"
+                  <> " is " <> previousVersion <> ")."
+          | otherwise ->
+              traverse_ (fst . snd) neededMigrations *> postMigrate $> MRMigrated previousVersion
+        Just dg
+          | previousVersion == dgoTargetVersion dg ->
+              pure MRNothingToDo
+          | otherwise -> 
+              case neededDownMigrations (dgoTargetVersion dg) of
+                Left reason -> 
+                  throw400 NotSupported $
+                    "This downgrade path (from "
+                      <> previousVersion <> " to " 
+                      <> dgoTargetVersion dg <> 
+                      ") is unfortunately not supported, because "
```

Comment:
Minor nit, but the word “unfortunately” here seems like a slightly odd word choice if the issue could be user error (e.g. specifying a nonexistent version). I think it would be fine to just drop it from the message, but I don’t feel super strongly about it.